openedx/features/content_type_gating/partitions.py

```diff
@@ -0,0 +1,220 @@
+    get_course_masquerade,
+    is_masquerading_as_specific_student,
+    get_masquerading_user_group,
+)
+from xmodule.partitions.partitions import Group, UserPartition, UserPartitionError
+
+
+LOG = logging.getLogger(__name__)
+
+CONTENT_GATING_PARTITION_ID = 51
+
+
+def create_content_gating_partition(course):
+    """
+    Create and return the Content Gating user partition.
+    """
+
+    try:
+        content_gate_scheme = UserPartition.get_scheme("content_type_gate")
+    except UserPartitionError:
+        LOG.warning("No 'content_type_gate' scheme registered, ContentTypeGatingPartitionScheme will not be created.")
+        return None
+
+    used_ids = set(p.id for p in course.user_partitions)
+    if CONTENT_GATING_PARTITION_ID in used_ids:
+        LOG.warning(
+            "Can't add 'content_type_gate' partition, as ID {id} is assigned to {partition} in course {course}.".format(
+                id=CONTENT_GATING_PARTITION_ID,
+                partition=_get_partition_from_id(course.user_partitions, CONTENT_GATING_PARTITION_ID).name,
+                course=unicode(course.id)
+            )
+        )
+        return None
+
+    partition = content_gate_scheme.create_user_partition(
+        id=CONTENT_GATING_PARTITION_ID,
+        name=_(u"Feature-based Enrollments"),
+        description=_(u"Partition for segmenting users by access to gated content types"),
+        parameters={"course_id": unicode(course.id)}
+    )
+    return partition
+
+
+class ContentTypeGatingPartition(UserPartition):
+    def access_denied_fragment(self, block, user, user_group, allowed_groups):
+        ecomm_service = EcommerceService()
+        ecommerce_checkout = ecomm_service.is_enabled(user)
+        ecommerce_checkout_link = ''
+        ecommerce_bulk_checkout_link = ''
+        verified_mode = None
+        CourseMode = apps.get_model('course_modes.CourseMode')
+        modes = CourseMode.modes_for_course_dict(block.scope_ids.usage_id.course_key)
+        verified_mode = modes.get(CourseMode.VERIFIED, '')
+
+        if ecommerce_checkout and verified_mode.sku:
+            ecommerce_checkout_link = ecomm_service.get_checkout_page_url(verified_mode.sku)
+
+        if verified_mode is None:
+            return None
+
+        request = crum.get_current_request()
+        if 'org.edx.mobile' in request.META.get('HTTP_USER_AGENT', ''):
+            upsell = ''
+        else:
+            upsell = textwrap.dedent("""\
+                <span class="certDIV_1" style="">
+                    <a href="{ecommerce_checkout_link}" class="certA_2">
+                        Upgrade to unlock  (${min_price})
+                    </a>
+                </span>
+            """.format(
+                ecommerce_checkout_link=ecommerce_checkout_link,
+                # TODO: Does this need i18n?
+                min_price=verified_mode.min_price,
+            ))
+
+        frag = Fragment(textwrap.dedent(u"""\
+            <div class=".content-paywall">
```

Comment:
could we use a template for this HTML?

Reply:
https://openedx.atlassian.net/browse/REV-569